### PR TITLE
feat(drive)!: allow using only one index in a query

### PR DESCRIPTION
**Pull request by shuplenkov**
## Issue being fixed or feature implemented

Since GroveDB doesn’t support indexes intersection, we need to validate query to make sure that only one index is being used. Also, there are some additional constraints due to the current GroveDB design. Potentially, they will be removed in the future.

## What was done?

- where conditions should follow a specific secondary index.
- ranges can be used only in the last where condition.
- sorting is allowed only for the last where condition and only for ranges.
- `elementMatch`, `contains` and `includes` temporary disabled

## How Has This Been Tested?

With unit and integration tests

## Breaking Changes

- where conditions should follow a specific secondary index
- ranges can be used only in the last where condition
- sorting is allowed only for the last where condition and only for ranges
- `elementMatch`, `contains` and `includes` temporary disabled

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/js-drive/lib/document/query/findNotIndexedFields.js [hunk 1]
```diff
@@ -21,21 +21,30 @@ function findNotIndexedFields(dataContractIndexFields, conditions) {
...
+        .find((index, i) => index
           // search through compound index
-          .find((element, i) => {
+          .find((element, k) => {
             const [indexField] = Object.keys(element);
             if (indexField !== field) {
               return false;
             }
 
+            if (indexToUse === undefined) {
```
Comment (shumkov) [inline packages/js-drive/lib/document/query/findNotIndexedFields.js:39]: Better use index names to compare

File: packages/js-drive/lib/document/query/findNotIndexedFields.js [hunk 2]
```diff
@@ -34,6 +36,13 @@ function findNotIndexedFields(dataContractIndexFields, conditions) {
...
+              indexToUse = index;
+            } else if (indexToUse !== index) {
+              // another index has already been used
+              return false;
```
Comment (shumkov) [inline packages/js-drive/lib/document/query/findNotIndexedFields.js:43]: User will a consfusing error that field is not indexed.

**Review by shumkov — CHANGES_REQUESTED** [review]
We need to validate `orderBy` too

File: packages/js-drive/lib/document/query/validateIndexedProperties.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            const [indexField] = Object.keys(element);
+            if (indexField !== field) {
+              return false;
+            }
+
+            if (indexToUse === undefined) {
+              indexToUse = index;
+            } else if (indexToUse !== index) {
+              // another index has already been used
+              result.addError(new FieldsFromMultipleIndicesError(field));
```
Comment (shumkov) [inline packages/js-drive/lib/document/query/validateIndexedProperties.js:54]: Shouldn't we return from the function here?

**Comment by shumkov** [comment]
@shuplenkov we got more clarity on how this thing should work:

All properties in `where` conditions must much exactly one of the indices and sequence of fields used in where must match how they defined in the index.

In `orderBy`, it can be only one property from the last condition

**Review by jawid-h — APPROVED** [review]
Looks legit

**Review by shumkov — APPROVED** [review]
👍

**Pull request merged**